src/qd8-f32-qb4w-gemm/gen/qd8-f32-qb4w-gemm-1x4c8-minmax-sse2-ld128.c

```diff
@@ -0,0 +1,192 @@
+        const __m128i vsb01c1 = _mm_cmpgt_epi8(_mm_setzero_si128(), vb01c1);
+        const __m128i vxb0c1 = _mm_unpacklo_epi8(vb01c1, vsb01c1);
+        const __m128i vxb1c1 = _mm_unpackhi_epi8(vb01c1, vsb01c1);
+
+        vacc0x0 = _mm_add_epi32(vacc0x0, _mm_madd_epi16(vxa0c1, vxb0c1));
+        vacc0x1 = _mm_add_epi32(vacc0x1, _mm_madd_epi16(vxa0c1, vxb1c1));
+        const __m128i vb23c1 = _mm_and_si128(vb23c01, vmask);
+        const __m128i vsb23c1 = _mm_cmpgt_epi8(_mm_setzero_si128(), vb23c1);
+        const __m128i vxb2c1 = _mm_unpacklo_epi8(vb23c1, vsb23c1);
+        const __m128i vxb3c1 = _mm_unpackhi_epi8(vb23c1, vsb23c1);
+
+        vacc0x2 = _mm_add_epi32(vacc0x2, _mm_madd_epi16(vxa0c1, vxb2c1));
+        vacc0x3 = _mm_add_epi32(vacc0x3, _mm_madd_epi16(vxa0c1, vxb3c1));
+
+
+        w = (const int8_t*) w + 32;
+        k -= 16 * sizeof(int8_t);
+      }
+
+      while (k >= 8 * sizeof(int8_t)) {
+        const __m128i va0 = _mm_loadl_epi64((const __m128i*) a0);
+        const __m128i vxa0 = _mm_srai_epi16(_mm_unpacklo_epi8(va0, va0), 8);
+        a0 += 8;
+
+        __m128i vb01 = _mm_load_si128((const __m128i*) w);
+        vb01 = _mm_slli_epi32(vb01, 4);
+        vb01 = _mm_and_si128(vb01, vmask);
+
+        const __m128i vsb01 = _mm_cmpgt_epi8(_mm_setzero_si128(), vb01);
+        const __m128i vxb0 = _mm_unpacklo_epi8(vb01, vsb01);
+        const __m128i vxb1 = _mm_unpackhi_epi8(vb01, vsb01);
+
+        vacc0x0 = _mm_add_epi32(vacc0x0, _mm_madd_epi16(vxa0, vxb0));
+        vacc0x1 = _mm_add_epi32(vacc0x1, _mm_madd_epi16(vxa0, vxb1));
+        __m128i vb23 = _mm_load_si128((const __m128i*) ((const int8_t*) w + 16));
+        vb23 = _mm_slli_epi32(vb23, 4);
+        vb23 = _mm_and_si128(vb23, vmask);
+
+        const __m128i vsb23 = _mm_cmpgt_epi8(_mm_setzero_si128(), vb23);
+        const __m128i vxb2 = _mm_unpacklo_epi8(vb23, vsb23);
+        const __m128i vxb3 = _mm_unpackhi_epi8(vb23, vsb23);
+
+        vacc0x2 = _mm_add_epi32(vacc0x2, _mm_madd_epi16(vxa0, vxb2));
+        vacc0x3 = _mm_add_epi32(vacc0x3, _mm_madd_epi16(vxa0, vxb3));
+
+        w = (const int8_t*) w + 32;
+        k -= 8 * sizeof(int8_t);
+      }
+      // accumulate float
+      const __m128 vfilter_output_scale0123 = _mm_load_ps((const float*) w);
+      w = (const float*) w + 4;
+
+      const __m128i vacc0x02 = _mm_add_epi32(_mm_unpacklo_epi32(vacc0x0, vacc0x2), _mm_unpackhi_epi32(vacc0x0, vacc0x2));
+      const __m128i vacc0x13 = _mm_add_epi32(_mm_unpacklo_epi32(vacc0x1, vacc0x3), _mm_unpackhi_epi32(vacc0x1, vacc0x3));
+
+      __m128i vacc0x0123 = _mm_add_epi32(_mm_unpacklo_epi32(vacc0x02, vacc0x13), _mm_unpackhi_epi32(vacc0x02, vacc0x13));
+
+      vout0x0123 = _mm_add_ps(vout0x0123, _mm_mul_ps(_mm_cvtepi32_ps(vacc0x0123), vfilter_output_scale0123));
+    }
+    __m128 one_sixteenth = _mm_set_ps1(1.0f/16);
+    vout0x0123 = _mm_mul_ps(vout0x0123, one_sixteenth);
+
+    const __m128 vinput_scale0 = _mm_load1_ps(&quantization_params[0].inv_scale);
```

Comment:
Possible to trade some `mul_ps` with a scalar one?
```suggestion
    const __m128 vinput_scale0 = _mm_set1_ps(quantization_params[0].inv_scale * (1/16.0));
```

Reply:
there is some special logic loading the vinput_scale from the struct, when MR > 1. Specifically we can get two inv_scales per load. I guess specifically for MR=1 we can do this, but not sure for MR > 1 this would give us any better perf. Any thoughts on that?

Reply:
So two things, 
* scalar pipes should be relatively open at this point, and we don't have any dependency, so this should be better than competing for the busy vector pipe for the vector multiplication
* That said, as you might know, I was doing some numerical accuracy analysis, and I feel if we move the 1/16.0f to weight (block) scales, then it might be better for both performance as well as accuracy given we don't have to do any multiplication in the kernel and the fp32 accumulation might be better with smaller number. To be clear, when convert larger int32 accumulator into fp32 in the first step we may drop some precision if the int32 is too large, however then multiplying with a smaller scale and accumulating in fp32 might still help). Let's chat offline? And I will update the summary here.

Reply:
Here you go #6596

Reply:
sure i'm ok with doing this 1/16 stuff as a follow up after this